Simple interfaces tutorial (#520)
ros2/ros2_documentation: source/Tutorials/Custom-ROS2-Interfaces.rst

It is good practice to keep ``.msg`` and ``.srv`` files in their own directories within a package.
Create the directories:

.. code-block:: console

  mkdir msg

  mkdir srv

2 Create custom definitions
^^^^^^^^^^^^^^^^^^^^^^^^^^^

2.1 msg definition
~~~~~~~~~~~~~~~~~~

In the ``interfaces/msg`` directory you just created, make a new file called ``Num.msg`` with one line of code declaring its data structure:

.. code-block:: console

    int64 num

This is your custom message that transfers a single integer called ``num``.

2.2 srv definition
~~~~~~~~~~~~~~~~~~

Back in the ``interfaces/srv`` directory you just created, make a new file called ``AddThreeInts.srv`` with the following request and response structure:

.. code-block:: console

  int64 a
  int64 b
  int64 c
  ---
  int64 sum

This is your custom service that requests three integers named ``a``, ``b``, and ``c``, and responds with an integer called ``sum``.

3 ``package.xml``
^^^^^^^^^^^^^^^^^

In order for your custom interfaces to be discoverable by other ROS 2 packages, add the following line to ``interfaces``’ ``package.xml`` file:

.. code-block:: xml

  <member_of_group>rosidl_interface_packages</member_of_group>

> .. MORE HERE? IT WORKS WHEN I JUST ADD THIS

Reviewer: In addition to the build_depend I mentioned in another comment, we should probably also add `<exec_depend>rosidl_default_runtime</exec_depend>` (like most of our packages do), though I'm not sure if it's strictly required. Maybe @dirk-thomas can shed some light?

Reviewer 2: > we should probably also add `<exec_depend>rosidl_default_runtime</exec_depend>`

Yes, that should certainly be done.

4 ``CMakeLists.txt``
^^^^^^^^^^^^^^^^^^^^

> Like ``pacakge.xml``, the ``CMakeLists.txt`` of any package with interfaces also needs some declarations in order for its interfaces to become discoverable.

Reviewer: Really, it's not just about the interfaces being discoverable. The code added to CMakeLists.txt converts the interfaces we defined into language-specific code (like C++ and Python) so that they can be used in those languages.

Add the following lines:

.. code-block:: console

>   find_package(rosidl_default_generators REQUIRED)

Reviewer: Because we rely on `rosidl_default_generators`  for generating language-specific code from the ROS interfaces, we should add a `build_depend` on it in the package.xml.

```
<build_depend>rosidl_default_generators</build_depend>
```

  rosidl_generate_interfaces(${PROJECT_NAME}
    "msg/Num.msg"
    "srv/AddThreeInts.srv"
   )

5 Build the ``interfaces`` package
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Now that all the parts of your custom interfaces package are in place, you can build the package.
In the root of your workspace (``~/dev_ws``), run the following command:

.. code-block:: console

    colcon build --packages-select interfaces

6 Confirm msg and srv creation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In a new terminal, run the following command from within your workspace (``dev_ws``) to source it:

.. code-block:: console

    . install/setup.bash

Now you can confirm that your interface creation worked by using the ``ros2 interface show`` command:

.. code-block:: console

    ros2 interface show interfaces/msg/Num

should return:

.. code-block:: console

    int64 num

And

.. code-block:: console

  ros2 interface show interfaces/srv/AddThreeInts

should return:

.. code-block:: console

    int64 a
    int64 b
    int64 c
    ---
    int64 sum

7 Test the new interfaces
^^^^^^^^^^^^^^^^^^^^^^^^^

For this step you can use the packages you created in previous tutorials.

> A few simple modifications in the nodes, ``CMakeLists`` and ``package`` files will allow you to see your interfaces in action.

Reviewer: ```suggestion
A few simple modifications in the ``CMakeLists.txt`` and ``package.xml`` files will allow you to see your interfaces in action.
```

7.1 Testing ``Num.msg`` with pub/sub
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> With some slight modifications to the publisher/subcriber package created in a previous tutorial, you can see ``Num.msg`` in action.

Reviewer: ```suggestion
With some slight modifications to the publisher/subscriber package created in a previous tutorial, you can see ``Num.msg`` in action.
```

                    'topic',
                    self.listener_callback,
                    10)
                self.subscription

            def listener_callback(self, msg):
                    self.get_logger().info('I heard: "%d"' % msg.num) # CHANGE

        def main(args=None):
            rclpy.init(args=args)

            minimal_subscriber = MinimalSubscriber()

            rclpy.spin(minimal_subscriber)

            minimal_subscriber.destroy_node()
            rclpy.shutdown()

        if __name__ == '__main__':
            main()

CMakeLists.txt:

Add the following lines (C++ only):

.. code-block:: console

    ...

    find_package(ament_cmake REQUIRED)
    find_package(rclcpp REQUIRED)
    find_package(interfaces REQUIRED)                         # CHANGE

    add_executable(talker src/publisher_member_function.cpp)
    ament_target_dependencies(talker rclcpp std_msgs)         # CHANGE

    add_executable(listener src/subscriber_member_function.cpp)
    ament_target_dependencies(listener rclcpp interfaces)     # CHANGE

    install(TARGETS
      talker
      listener
      DESTINATION lib/${PROJECT_NAME})

    ament_package()

> Package.xml:

Reviewer: ```suggestion
package.xml:
```

                            'Service call failed %r' % (e,))
                    else:
                        minimal_client.get_logger().info(
                            'Result of add_three_ints: for %d + %d + %d = %d' %                               # CHANGE
                            (minimal_client.req.a, minimal_client.req.b, minimal_client.req.c, response.sum)) # CHANGE
                    break

            minimal_client.destroy_node()
            rclpy.shutdown()

        if __name__ == '__main__':
            main()

CMakeLists.txt:

Add the following lines (C++ only):

.. code-block:: console

    ...

    find_package(ament_cmake REQUIRED)
    find_package(rclcpp REQUIRED)
    find_package(interfaces REQUIRED)        # CHANGE

    add_executable(server src/add_two_ints_server.cpp)
    ament_target_dependencies(server
      rclcpp interfaces)                      #CHANGE

    add_executable(client src/add_two_ints_client.cpp)
    ament_target_dependencies(client
      rclcpp interfaces)                      #CHANGE

    install(TARGETS
      server
      client
      DESTINATION lib/${PROJECT_NAME})

    ament_package()

> Package.xml:

Reviewer: ```suggestion
package.xml:
```
